### PR TITLE
Parameterise YuGCN

### PR DESCRIPTION
**Pull request by htwangtw**
As title.

@ltetrel  this should be ready to merge. I will do separate PRs to add non-guided back prop and deal with typing.

### CODE REVIEW

File: gcn_package/models/gcn.py [hunk 1]
```diff
@@ -34,8 +36,8 @@ def forward(self, x):
         x = F.relu(x)
         x = self.conv6(x, self.edge_index, self.edge_weight)
         x = tg.nn.global_mean_pool(x,torch.from_numpy(np.array(range(x.size(0)),dtype=int)))
-        
-        x = x.view(-1, 512*32)
+
+        x = x.view(-1, self.batch_size * self.n_roi)
```
Comment (ltetrel) [inline gcn_package/models/gcn.py:40]: batch_size is confusing here. That does not make sense that the model depends on data generator batch size, it should be independent.
The best way to flatten the actual output is to simply use:
x = torch.flatten(x, 1) (please double check dimension here)
This way batch_size is not needed

File: gcn_package/models/gcn.py [hunk 2]
```diff
@@ -6,22 +6,24 @@
...
         self.edge_weight = edge_weight
+        self.batch_size, self.n_roi = batch_size, n_roi
         self.conv1 = tg.nn.ChebConv(in_channels=n_timepoints,out_channels=32,K=2,bias=True)
         self.conv2 = tg.nn.ChebConv(in_channels=32,out_channels=32,K=2,bias=True)
         self.conv3 = tg.nn.ChebConv(in_channels=32,out_channels=32,K=2,bias=True)
         self.conv4 = tg.nn.ChebConv(in_channels=32,out_channels=32,K=2,bias=True)
         self.conv5 = tg.nn.ChebConv(in_channels=32,out_channels=32,K=2,bias=True)
         self.conv6 = tg.nn.ChebConv(in_channels=32,out_channels=32,K=2,bias=True)
-        self.fc1 = nn.Linear(512*32, 256)
+        self.fc1 = nn.Linear(batch_size * n_roi, 256)
```
Comment (ltetrel) [inline gcn_package/models/gcn.py:20]: This is weird that linear needs batch_size.
what should be important is the dimension of the feature (ex for an image 256x256, for ts n_roi x time)
A forward pass is done on just one sample at a time (a single ts), not multiple at a time. Then the gradient is ~averaged over al samples from the batch.
Or is it specific from Yu's model ?

**Review by ltetrel — CHANGES_REQUESTED** [review]
Overall needs clarification on why batch size is a model parameters.
One model pass should not depend on batch size.
(I know that this was there before, but I did not really check the model spec before)

File: gcn_package/models/gcn.py (hunk 1 above)
Comment (htwangtw) [reply gcn_package/models/gcn.py:40]: I can modify this line to what you suggested here

File: gcn_package/models/gcn.py (hunk 2 above)
Comment (htwangtw) [reply gcn_package/models/gcn.py:20]: I simply found out this number is consistently the number of batch_size * n_roi, as it need to be the same as line 40

Comment (htwangtw) [reply gcn_package/models/gcn.py:20]: If we revert this back to Yu's original code, and implement your suggestion for Line 40, this is what we got:

```
RuntimeError: mat1 and mat2 shapes cannot be multiplied (32x3648 and 16384x256)
```

**Comment by ltetrel** [comment]
Can you test now with my latest addition ?

**Comment by htwangtw** [comment]
Yarp it works!

File: gcn_package/models/gcn.py [hunk 3]
```diff
@@ -36,7 +36,7 @@ def forward(self, x):
         x = self.conv6(x, self.edge_index, self.edge_weight)
         x = tg.nn.global_mean_pool(x, torch.from_numpy(np.array(range(x.size(0)), dtype=int)))
 
-        x = torch.nn.Flatten()
+        x = torch.flatten(x, 1)
```
[the remark this answers is not in the document]
Comment (ltetrel) [reply gcn_package/models/gcn.py:39]: Oups sorry I confused myself with the tensorflow functions. Indeed second parameter specify which start dim to flatten (in our case dim1, dim0 is the batch dim)

**Pull request merged**